File: LeetCode/Study Plan/Practice Problems/Medium/Implement Trie/Go/implement_trie.go
package main

type TrieNode struct {
	is_word bool
	children [26]*TrieNode
}

type Trie struct {
	root *TrieNode    
}


func Constructor() Trie { return Trie{root : &TrieNode{}} }


func (this *Trie) Insert(word string)  {
    root := this.root
    for _, ch := range word {
    	index := ch - 'a'
    	if root.children[index] == nil {
    		root.children[index] = &TrieNode{}
    	}
    	root = root.children[index]
    }
   	root.is_word = true
}


func (this *Trie) Search(word string) bool {
    root := this.root
    for _,ch := range word {
    	index := ch - 'a'
    	if root.children[index] == nil {
    		return false
    	}
    	root = root.children[index]
    }
    return root.is_word
}


func (this *Trie) StartsWith(prefix string) bool {
	root := this.root
	for _, ch := range prefix {
		index := ch - 'a'
		if root.children[index] == nil {
			return false
		}
		root = root.children[index]
	}
	return true
}


func main() {

}

File: LeetCode/Study Plan/Practice Problems/Medium/Implement Trie/Java/implementTrie.java
class TrieNode {
    TrieNode[] children;
    Boolean isWord;
    public TrieNode() {
        children = new TrieNode[26];
        isWord = false;
    }
}

class Trie {
    
    TrieNode root;
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode r = this.root;
        for (char c : word.toCharArray()) {
            if (r.children[c-'a'] == null) {
                r.children[c-'a'] = new TrieNode();
            }
            r = r.children[c-'a'];
        }
        r.isWord = true;
    }
    
    public boolean search(String word) {
        TrieNode r = this.root;
        for (char c : word.toCharArray()) {
            if (r.children[c-'a'] == null) {
                return false;
            }
            r = r.children[c-'a'];
        }
        return r.isWord ? true : false;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode r = this.root;
        for (char c : prefix.toCharArray()) {
            if (r.children[c-'a'] == null) {
                return false;
            }
            r = r.children[c-'a'];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
